[RFC] Add functions for dumping and restoring msgpack values

ZyX-I opened the pull request:
After second commit the result is much more precise then what you can obtain by converting msgpack to Object and then to typval_T using standard converters.

New functions are needed for the plugin that reads and writes ShaDa files (#2506). Maybe they will be needed for other files in the future because e.g. persistent undo files are rather not extensible and rewriting the format with messagepack is easier then creating our own extensible binary format from scratch.

Note about second commit: I am using recursive conversion of messagepack values to typval_T because messagepack parsing result cannot possibly be self-referencing. If needed I can make this non-recursive as well. I am using non-recursive stack-based loop for typval_T to messagepack conversion because typval_T values _can_ reference itself and I do not like the idea of limiting conversion by some random value with reason “recursion depth exceeded”, direct checking whether container is self-referencing is much better.

kvec_t is used because garray_T does not support `kv_pop` and I do not know any other structures in NeoVim source code which may fit in. But this means one realloc for each kv_push which increases stack size, garray_T may have done less reallocs. Do not think this will cause performance problems for the intended usage, especially with faster allocator.

ZyX-I committed:
ex_cmds*: Silence -Wstrict-prototypes warnings

Due to a number of places where headers generated from these files are included
gcc with -DEXITFREE produces *loads* of warnings for these functions.

ZyX-I committed:
eval: Add msgpackparse and msgpackdump functions

ZyX-I committed:
eval: Remove most of msgpack* functions limitations

justinmk commented on src/nvim/eval.c:
```diff
@@ -11841,6 +11931,830 @@ static void f_mode(typval_T *argvars, typval_T *rettv)
+  FUNC_ATTR_NONNULL_ALL
+{
+  if (argvars[0].v_type != VAR_LIST) {
+    EMSG2(_(e_listarg), "msgpackdump()");
```
typo

justinmk commented on src/nvim/eval.c:
```diff
@@ -11841,6 +11931,830 @@ static void f_mode(typval_T *argvars, typval_T *rettv)
 }
 
+/// Msgpack callback for writing to readfile()-style list
+static int msgpack_list_write(void *data, const char *buf, size_t len)
```
Perhaps we should create a `src/nvim/eval/` subdirectory that contains eval sub-modules like this, and this could live in `eval/msgpack.c`. Or `eval_msgpack.c` at the top level is fine too.

ZyX-I commented:
Build .5 (MSAN) fails due to an unrelated issue:

```
==7794== WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7f4b8f3a44a6 in os_isdir /home/travis/build/neovim/neovim/src/nvim/os/fs.c:69:7
    #1 0x7f4b8e951aa4 in vim_maketempdir /home/travis/build/neovim/neovim/src/nvim/tempfile.c:36:10
    #2 0x7f4b8e951353 in vim_gettempdir /home/travis/build/neovim/neovim/src/nvim/tempfile.c:90:5
    #3 0x7f4b8e952972 in vim_tempname /home/travis/build/neovim/neovim/src/nvim/tempfile.c:127:21
    #4 0x7f4b8f50f61e in server_init /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/server.c:38:30
    #5 0x7f4b8cf23162 in event_init /home/travis/build/neovim/neovim/src/nvim/main.c:154:3
    #6 0x7f4b8cf270bb in main /home/travis/build/neovim/neovim/src/nvim/main.c:278:3
    #7 0x7f4b8a94776c in __libc_start_main /build/buildd/eglibc-2.15/csu/libc-start.c:226
    #8 0x7f4b8bf10aa0 in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0xaeaa0)

  Uninitialized value was created by an allocation of 'request' in the stack frame of function 'os_stat'
    #0 0x7f4b8f3aa120 in os_stat /home/travis/build/neovim/neovim/src/nvim/os/fs.c:207

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/travis/build/neovim/neovim/src/nvim/os/fs.c:69 os_isdir
Exiting
```

. I did not modify any functions in this stack in this PR.

ZyX-I commented:
.8 (MINGW) errors for even more unrelated reason:

```
$ .ci/install.sh
Downloading MinGW...
xz: (stdin): File format not recognized
tar: Child returned status 1
tar: Error is not recoverable: exiting now

The command ".ci/install.sh" failed and exited with 2 during .
```

ZyX-I committed:
vim_to_msgpack: Silence -Werror=unitialized compiler error

Compiler warning was found in [QuickBuild logs][1] from [this page][2]. GCC and
clang on travis appear to be fine. Relevant log parts:

    Step Log (master>buildall>build-node?testNode=linux-64>build-and-run-tests>build-and-run-tests-parameterized?buildType=Release>configure-neovim-and-build-nvim)
    <…>
    16:26:31,364 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c: In function ‘f_msgpackdump’:
    16:26:31,364 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c:12371:26: error: ‘cur_tv’ may be used uninitialized in this function [-Werror=uninitialized]
    16:26:31,364 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval.c:12328:21: note: ‘cur_tv’ was declared here
    16:26:31,938 WARN  - cc1: all warnings being treated as errors

[1]: http://neovim-qb.szakmeister.net/wicket/page?4-1.ILinkListener-content-buildTab-panel-errorContainer-steps-5-logLink
[2]: http://neovim-qb.szakmeister.net/build/2099/overview

justinmk commented:
The mingw issue is because sourceforge is a piece of shit.

splinterofchaos commented on runtime/doc/eval.txt:
```diff
@@ -4613,6 +4621,90 @@ mode([expr])	Return a string that indicates the current mode.
+		1. Mapping ordering is not preserved unless messagepack 
+		   mapping is dumped using generic  mapping 
+		   (|msgpack-special-map|).
+		2. Since parser aims to preserve all data untouched (except 
```
Grammar: "Since *the* parser"

splinterofchaos commented on src/nvim/eval.c:
```diff
@@ -5177,6 +5265,23 @@ void list_append_string(list_T *l, char_u *str, int len)
+/// @param[out]  l    List to append to.
+/// @param[in]   str  String to append.
+void list_append_allocated_string(list_T *l, char *const str)
+  FUNC_ATTR_NONNULL_ARG(1)
```
Oh, I've wanted this often. :+1:

justinmk replied:
`list_append_string` can call `list_append_allocated_string()` instead of repeating the logic.

ZyX-I commented:
> The mingw issue is because sourceforge is a piece of shit.

How about using Gentoo distfiles mirrors? E.g. http://mirror.yandex.ru/gentoo-distfiles/distfiles/mingw-w64-v4.0.1.tar.bz2 (or http://mirror.yandex.ru/gentoo-distfiles/distfiles/mingwrt-4.0.3-1-mingw32-src.tar.lzma, I do not really know what is the proper source package, maybe more are needed: for cross-compiling Gentoo has a number of packages: binutils, runtime, compiler, gdb, w32api (headers)). You will have to build this by bot though.

Also I am pretty sure there is a PPA for this; most likely the result will work if you unpack .deb like regular archive and set some environment variables.

splinterofchaos commented on src/nvim/eval.c:
```diff
@@ -11841,6 +11948,830 @@ static void f_mode(typval_T *argvars, typval_T *rettv)
+    line_end = memchr(line_start, NL, end - line_start);
+    if (line_end == NULL) {
+      line_end = end;
+    }
```
`xmemscan` would return `line_end` instead of `NULL`.

ZyX-I commented:
Just realized that functions used to convert to `typval_T` directly would be more convenient and flexible (regarding forward compatibility) for use in #2506 then functions for converting to some of the API structures like Dictionary or Array.

Thus I would say this PR is blocking #2506.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -4613,6 +4621,91 @@ mode([expr])	Return a string that indicates the current mode.
+		   use.
+							*msgpack-special-dict*
+		Some messagepack strings may be parsed to a special 
+		dictionaries. Special dictionaries are dictionaries which
```
"to a special" => "to special"

ZyX-I replied:
Fixed.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -4613,6 +4621,91 @@ mode([expr])	Return a string that indicates the current mode.
+msgpackparse({list})					*msgpackparse()*
+		Convert a |readfile()|-style list to a list of VimL objects. 
+		Example: >
+			let fname = expand('~/.nvim/shada/main.shada)
```
missing closing single-quote

ZyX-I replied:
Fixed.

ZyX-I committed:
documentation: Add {Nvim} tag to new functions

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -5157,24 +5245,40 @@ void list_append_dict(list_T *list, dict_T *dict)
+    list_append_allocated_string(l, NULL);
+  } else {
+    list_append_allocated_string(l, (len >= 0
+                                     ? xstrndup((char *) str, len)
```
Use `xmemdupz` here? I always use this variant for performance reasons when I am pretty sure that there are either no embedded zero bytes or that they do not matter.

justinmk replied:
Yes, that's better.

ZyX-I replied:
Done.

ZyX-I committed:
eval: Call list_append_allocated_string from list_append_string

justinmk commented on src/nvim/eval.c:
```diff
@@ -512,6 +580,26 @@ void eval_init(void)
+  }
+  msgpack_types_dict->dv_lock = VAR_FIXED;
+
+  set_vim_var_dict(VV_MSGPACK_TYPES, msgpack_types_dict);
```
move this block to a separate function, `eval_msgpack_init` ?

ZyX-I replied:
@justinmk I do not think that 20 lines are enough to move this out. `eval_init` is still not very lengthy.

justinmk commented on src/nvim/eval.c:
```diff
@@ -11841,6 +11947,827 @@ static void f_mode(typval_T *argvars, typval_T *rettv)
+///
+/// @return true in case of success, false in case of failure.
+static inline bool vim_list_to_buf(const list_T *const list,
+                                   size_t *const ret_len, char **const ret_buf)
```
I like this `ret_foo` convention for out params. Maybe we should make a habit of that.

justinmk commented on src/nvim/eval.c:
```diff
@@ -11841,6 +11947,827 @@ static void f_mode(typval_T *argvars, typval_T *rettv)
+  return 0;
+}
+
+/// Convert readfile()-style list to a char * buffer with length
```
@splinterofchaos Didn't you already write a function like this?

justinmk commented:
I need to step out for a few hours, but I won't hold this up. The user-facing API looks good, and the tests look good. If no one has any comments later today, I will merge.

justinmk commented on test/functional/viml/msgpack_functions_spec.lua:
```diff
@@ -0,0 +1,606 @@
+local helpers = require('test.functional.helpers')
```
FYI there is now a `test/functional/eval/` directory, I suggest putting this there. I will move the other resident of this `viml/` directory somewhere else.

ZyX-I committed:
functests: Use eval subdirectory in place of viml

fwalch commented on runtime/doc/eval.txt:
```diff
@@ -4613,6 +4621,91 @@ mode([expr])	Return a string that indicates the current mode.
+			let fname = expand('~/.nvim/shada/main.shada')
+			let mpack = readfile(fname, 'b')
+			let shada_objects = msgpackparse(mpack)
+<		This will read |shada-file| to `shada_objects` list.
```
The tag `*shada-file*` is only introduced in #2506 and doesn't exist in `master`. As #2506 is scheduled for 0.2, this example should maybe be changed for 0.1.

justinmk replied:
This can be an anti-easter-egg until shada gets merged :sweat_smile:
